core/src/main/java/org/apache/accumulo/core/data/Mutation.java

```diff
@@ -306,6 +318,8 @@ private void put(byte[] cf, int cfLength, byte[] cq, int cqLength, byte[] cv, bo
     if (buffer == null) {
       throw new IllegalStateException("Can not add to mutation after serializing it");
     }
+    estimatedSize += cfLength + cqLength + (hasts ? 8 : 0) + valLength + 2 * 1 + 4 * SERIALIZATION_OVERHEAD;
+    Preconditions.checkArgument(estimatedSize < MAX_MUTATION_SIZE && estimatedSize >= 0, "Maximum mutation size must be less than 2GB ");
```

Comment:
I was thinking of a slightly different approach that will give more accurate size estimates.  Something like the following.

 * Use estimatedSize to track row and large values only.
 * Add size() method to `UnsynchronizedBuffer.Writer` that returns its offset.  Then can call `buffer.size()` to get an accurate count of amount of data written.

Then this check could be something like

```java
  long estSizeAfterPut = estimatedSize + buffer.size() + cfLength + cqLength + (hasts ? 8 : 0) + valLength + 2 * 1 + 4 * SERIALIZATION_OVERHEAD;
Preconditions.checkArgument(estSizeAfterPut < MAX_MUTATION_SIZE && estSizeAfterPut >= 0, "Maximum mutation size must be less than 2GB ");
```